### PR TITLE
[Join] Refactored the code of join and support left/right semi anti join and other join conjunt

### PR DESCRIPTION
**Pull request by HappenLee**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: gensrc/script/doris_builtins_functions.py [hunk 1]
```diff
@@ -570,7 +570,7 @@
             '_ZN5doris13MathFunctions5roundEPN9doris_udf15FunctionContextERKNS1_9DoubleValE', '', '', 'vec', ''],
     [['round', 'dround'], 'DOUBLE', ['DOUBLE', 'INT'],
             '_ZN5doris13MathFunctions11round_up_toEPN9doris_udf'
-            '15FunctionContextERKNS1_9DoubleValERKNS1_6IntValE', '', '', 'vec', ''],
+            '15FunctionContextERKNS1_9DoubleValERKNS1_6IntValE', '', '', '', ''],
```
Comment (wangbo) [inline gensrc/script/doris_builtins_functions.py:573]: 这个地方怎么把round函数的向量化版给删了

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 2]
```diff
@@ -95,7 +109,11 @@ struct ProcessHashTableProbe {
               _arena(join_node->_arena),
               _rows_returned_counter(join_node->_rows_returned_counter) {}
 
-    Status operator()(HashTableContext& hash_table_ctx, ConstNullMapPtr null_map,
+    // Only process the join with no other join conjunt, because of no other join conjunt
+    // the output block strcut is same with mutable block. we can do more opt on it and simplify
```
Comment (wangbo) [inline be/src/vec/exec/join/vhash_join_node.cpp:113]:
```suggestion
    // the output block struct is same with mutable block. we can do more opt on it and simplify
```

File: be/src/vec/exec/join/join_op.h [hunk 3]
```diff
@@ -28,9 +28,12 @@ struct RowRef {
 
     const Block* block = nullptr;
     SizeT row_num = 0;
+    // Use in right join to mark row is visited
+    // TODO: opt the varaible to use it only need
+    bool visited = false;
 
     RowRef() {}
-    RowRef(const Block* block_, size_t row_num_) : block(block_), row_num(row_num_) {}
+    RowRef(const Block* block_, size_t row_num_, bool visited_ = false) : block(block_), row_num(row_num_), visited(visited_) {}
```
Comment (yangzhg) [inline be/src/vec/exec/join/join_op.h:36]: Why add `_` at the end?

File: gensrc/script/doris_builtins_functions.py (hunk 1 above)
Comment (HappenLee) [reply gensrc/script/doris_builtins_functions.py:573]: 这个是之前测试跑回归的时候这个函数有bug，会core，所以删除了。我再加回去

File: be/src/vec/exec/join/join_op.h (hunk 3 above)
Comment (HappenLee) [reply be/src/vec/exec/join/join_op.h:36]: 这个是从clickhouse移过来的代码，我统一改成我们的风格把

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 4]
```diff
@@ -106,6 +124,102 @@ struct ProcessHashTableProbe {
...
+                                auto &column = *it->block->get_by_position(j).column;
+                                mcol[j + right_col_idx]->insert_from(column, it->row_num);
+                            }
+                        }
+                        it->visited = true;
+                    }
+                }
+            } else if (_join_node->_match_all_probe || _join_node->_join_op == TJoinOp::LEFT_ANTI_JOIN) {
+                ++current_offset;
+                // only full outer / left outer need insert the data of right table
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/exec/join/vhash_join_node.cpp:171]: anti join 和 semi join都只输出一边

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 5]
```diff
@@ -154,73 +285,132 @@ struct ProcessHashTableProbe {
...
+
+                    if (!other_hit) {
+                        for (size_t j = 0; j < right_col_len; ++j) {
+                            typeid_cast<ColumnNullable*>(std::move(*output_block->get_by_position(j + right_col_idx).column).mutate().get())
+                                    ->get_null_map_data()[i] = true;
+                        }
+                    }
+
+                    if (join_hit) {
+                        if (visited_map[i]) *visited_map[i] |= other_hit;
```
Comment (wangbo) [inline be/src/vec/exec/join/vhash_join_node.cpp:316]: 这行应该没啥用吧，如果visited_map[i]是true，那么它和任何值做or，结果应该也都还是true啊

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 6]
```diff
@@ -154,73 +285,134 @@ struct ProcessHashTableProbe {
...
+                        // if not, only some key must keep one row will output will null right table column
+                        if (same_to_prev[i] && filter_map.back() && !column->get_bool(i - 1)) filter_map[i - 1] = false;
+                    } else {
+                        filter_map.push_back(true);
+                    }
+                }
+                output_block->get_by_position(result_column_id).column = std::move(new_filter_column);
+            } else if (_join_node->_join_op == TJoinOp::RIGHT_OUTER_JOIN) {
+                for (int i = 0; i < column->size(); ++i) {
+                    if (visited_map[i]) *visited_map[i] |= column->get_bool(i);
```
Comment (wangbo) [inline be/src/vec/exec/join/vhash_join_node.cpp:329]:
```suggestion
                    *visited_map[i] |= column->get_bool(i);
```

**Review by wangbo — APPROVED** [review]
LGTM

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 7]
```diff
@@ -154,73 +285,135 @@ struct ProcessHashTableProbe {
...
+                    if ((!visited_map[i] || !column->get_bool(i))) {
+                        filter_map.push_back(true);
+                        auto loc = i;
+                        if (same_to_prev[loc]) {
+                            do {
+                                filter_map[--loc] = false;
+                            } while (same_to_prev[loc]);
+                        }
+                    }
+                    else
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/exec/join/vhash_join_node.cpp:358]: ok, 辛苦了

**Pull request merged**